src/common/src/mlib/ckdint.h

```diff
@@ -0,0 +1,672 @@
+            } else if (a_is_negative) {
+               if (sum_is_large) {
+                  // Same as above case with sum_is_large
+                  return true;
+               }
+            }
+            return false;
+         } else { // U = S + U
+            return (intmax_t) ((sum ^ a ^ signbit) & (sum ^ b)) < 0;
+            // Expanded:
+            const bool sum_is_large = sum > INTMAX_MAX;
+            const bool b_is_large = b > INTMAX_MAX;
+            const bool a_is_negative = (intmax_t) a < 0;
+
+            if (!a_is_negative && b_is_large) {
+               // We are adding a non-negative value to a large number, so the
+               // sum must also be large
+               if (!sum_is_large) {
+                  // We ended up with a smaller value, meaning that we must have wrapped
+                  return true;
+               }
+            }
+            if (a_is_negative && !b_is_large) {
+               // We subtracted a non-negative value from a non-large number, so
+               // the result should not be large
+               if (sum_is_large) {
+                  // We ended up with a large value, so we must have wrapped
+                  return true;
+               }
+            }
+            return false;
+         }
+      } else {
+         if (b_signed) { // U = U + S  --- (See [U = S + U] for an explanation)
+            return (intmax_t) ((sum ^ a) & (sum ^ b ^ signbit)) < 0;
+         } else { // U = U + U (simple case)
+            return sum < a;
+         }
+      }
+   }
+}
+
+// Support for the `mlib_sub` macro
+static inline bool (mlib_sub) (uintmax_t *dst, bool dst_signed, bool a_signed, uintmax_t a, bool b_signed, uintmax_t b)
+   mlib_noexcept
+{
+   // Perform the subtraction using regular wrapping arithmetic
+   const uintmax_t diff = *dst = a - b;
+   const uintmax_t signbit = (UINTMAX_C (1) << ((sizeof (intmax_t) * CHAR_BIT) - 1));
+   // Test whether the operation overflowed for the given sign configuration
+   // (See mlib_add for more details on why we do this bit fiddling)
+   if (dst_signed) {
+      const bool diff_is_negative = (intmax_t) diff < 0;
+      if (a_signed) {
+         if (b_signed) { // S = S - S
+            return (intmax_t) ((a ^ b) & (diff ^ a)) < 0;
+            // Explain:
+            const bool a_is_negative = (intmax_t) a < 0;
+            const bool b_is_negative = (intmax_t) b < 0;
+            if (a_is_negative != b_is_negative) {
+               // Given: Pos - Neg = Pos
+               //      ∧ Neg - Pos = Neg
+               // We expect that the difference preserves the sign as the minuend
+               if (diff_is_negative != a_is_negative) {
+                  return true;
+               }
+            }
+            // Otherwise, `Pos - Pos` and `Neg - Neg` cannot possibly overflow
+            return false;
+         } else { // S = S - U
+            // The diff overflows if the negated minuend is smaller than the subtrahend
```

Comment:
```suggestion
            // The diff overflows if the sign-bit-flipped minuend is smaller than the subtrahend
```

This logic looks right, but the word "negated" seems misleading. If I understand correctly, flipping the sign bit maps the signed value to a value that has the same ordering as unsigned values. For example, the ordering of signed values for 3 bits:
```
100 ... 111 000 001 ... 011
```
After flipping the first bit, the ordering matches the ordering for unsigned values:
```
000 ... 011 100 101 ... 111
```